update geometry/scanning (#101)
OI-wiki/i18n: docs/geometry/scanning.md

## Introduction

> Scanning lines are usually used on the graphics. It is very similar to its literal meaning, that is, a line sweeps across the entire map. It is generally used to solve questions like graphic area, perimeter and other.

Reviewer: 统一一下用词？下同
image？

Reviewer: ```suggestion
Scanning lines are usually used on the graphics. It is very similar to its literal meaning, that is, a line sweeps across the entire map. It is generally used to solve problems like areas, perimeters.
```

## Atlantis problem

### Description

In a two-dimensional coordinate system, given the lower left and upper right coordinates of multiple rectangles, find the area of the figure formed by all rectangles.

### Solution

According to the figure below, it can be seen that the total area can be obtained directly using brute force solution. But what if the input is too large? At this time, we need to consider using the **scanning line** algorithm.

### Process

Now suppose we have a line and start scanning from bottom to top:

![](./images/scanning-1-en.png)

![](./images/scanning-2-en.png)

![](./images/scanning-3-en.png)

![](./images/scanning-4-en.png)

![](./images/scanning-5-en.png)

![](./images/scanning-6-en.png)

![](./images/scanning-7-en.png)

> > The original version of above figures are made by [kk303](https://blog.csdn.net/kk303).

Reviewer: ```suggestion
> The original version of above figures were made by [kk303](https://blog.csdn.net/kk303).
```

- As shown in the figure, we can divide the entire rectangle into small rectangles with different colors. Then the height of this small rectangle is the distance we swept across. And then there is one variable left, that is, the length of the rectangle that keeps changing.

> - Our segment tree is used to maintain the length of the rectangle. We mark the upper and lower sides of each rectangle where the lower side is marked as $1$ , and the upper side is marked as $-1$ . Whenever we encounter a rectangle, we know the line that is marked as $1$ . We add the length of this rectangle, and when our scan reaches $-1$ , it means that this edge needs to be deleted, so we delete it. Using $1$ and $-1$ can easily reach this state.

Reviewer: ```suggestion
- The segment tree here is used to maintain the length of the rectangle. We mark the upper and lower sides of each rectangle in which the lower sides are marked as $1$ , while the upper sides are marked as $-1$ . Whenever we encounter a rectangle, we know the line that is marked as $1$ . We add the length of this rectangle, and when our scan reaches $-1$ , it means that this edge needs to be deleted, so did we. Using $1$ and $-1$ can easily reach this state.
```
原文 "的到" 是想说得到吗

> - Please note that the segment tree here does not mean an end point of a line segment, instead, but an interval. So what we have to calculate is $r+1$ and $r-1$ .

Reviewer: ```suggestion
- Please note that the segment tree here does not mean an end point of a line segment, but an interval instead. So what we have to calculate are $r+1$ and $r-1$ .
```

          s[i + 1] = y1;
          s[i + n + 1] = y2;
        }
        sort(s + 1, s + (2 * n + 1));  // discretization

>         sort(p, p + 2 * n, cmp);  // sort the coordinates of the sides of the rectangle ascendingly

Reviewer: 纵坐标

    }
    ```

## 练习
## Practice problems

-  [「HDU1542」Atlantis](http://acm.hdu.edu.cn/showproblem.php?pid=1542) 

-  [「HDU1828」Picture](http://acm.hdu.edu.cn/showproblem.php?pid=1828) 

-  [「HDU3265」Posters](http://acm.hdu.edu.cn/showproblem.php?pid=3265) 

## 参考资料

-  <https://www.cnblogs.com/yangsongyi/p/8378629.html> 

-  <https://blog.csdn.net/riba2534/article/details/76851233> 

> -  <https://blog.csdn.net/winddreams/article/details/38495093> 

Reviewer: reference 中文的也可以留吧
